rust/sword/src/jets/cold.rs

```diff
@@ -767,3 +767,650 @@ impl Cold {
+        // for battery in self {
+        //     reversed.push(battery);
+        // }
+        // reversed.reverse();
+        let mut list = D(0);
+        for (battery, parent_axis) in self {
+            let battery_noun = unsafe { *battery };
+            let parent_axis_noun = parent_axis.as_noun();
+            let item = T(stack, &[battery_noun, parent_axis_noun]);
+            list = T(stack, &[item, list]);
+        }
+        list
+    }
+
+    fn from_noun<A: NounAllocator>(stack: &mut A, noun: &Noun) -> NounableResult<Self::Target> {
+        let mut batteries = NO_BATTERIES;
+        for item in NounListIterator(noun.clone()) {
+            // FIXME: NotCell error here too
+            let cell = item.cell().ok_or(FromNounError::NotCell)?;
+            let battery = cell.head();
+            let parent_axis = cell.tail().as_atom()?;
+            let batteries_mem: *mut BatteriesMem = unsafe { stack.alloc_struct(1) };
+            unsafe {
+                batteries_mem.write(BatteriesMem {
+                    battery,
+                    parent_axis,
+                    parent_batteries: batteries,
+                });
+            }
+            batteries = Batteries(batteries_mem);
+        }
+        Ok(batteries)
+    }
+}
+
+impl Nounable for BatteriesList {
+    type Target = BatteriesList;
+    fn into_noun<A: NounAllocator>(self, stack: &mut A) -> Noun {
+        let mut list = D(0);
+        for batteries in self {
+            let batteries_noun = batteries.into_noun(stack);
+            list = T(stack, &[batteries_noun, list]);
+        }
+        list
+    }
+
+    fn from_noun<A: NounAllocator>(stack: &mut A, noun: &Noun) -> NounableResult<Self::Target> {
+        let mut batteries_list = BATTERIES_LIST_NIL;
+        for item in NounListIterator(noun.clone()) {
+            let batteries = Batteries::from_noun(stack, &item)?;
+            let batteries_list_mem: *mut BatteriesListMem = unsafe { stack.alloc_struct(1) };
+            unsafe {
+                batteries_list_mem.write(BatteriesListMem {
+                    batteries,
+                    next: batteries_list,
+                });
+            }
+            batteries_list = BatteriesList(batteries_list_mem);
+        }
+        Ok(batteries_list)
+    }
+}
+
+impl<T: Nounable + Copy + mem::Preserve> Nounable for Hamt<T> {
+    type Target = Vec<(Noun, T::Target)>;
+
+    // FIXME: This should be reversed for linked-list construction
+    fn into_noun<A: NounAllocator>(self, stack: &mut A) -> Noun {
+        let mut list = D(0);
+                let mut reverse = Vec::new();
+        for item in self.iter() {
+            reverse.push(item);
+        }
+        reverse.reverse();
+        for slice in reverse {
+            for (key, value) in slice {
+                let key_noun = key.into_noun(stack);
+                let value_noun = value.into_noun(stack);
+                let items = T(stack, &[key_noun, value_noun]);
```

Comment:
`key` is already a `Noun`, right?
```suggestion
                let value_noun = value.into_noun(stack);
                let items = T(stack, &[key, value_noun]);
```